src/backend/metal/src/command.rs

```diff
@@ -346,37 +445,58 @@ impl core::RawCommandBuffer<Backend> for CommandBuffer {
         unimplemented!()
     }
 
-    fn bind_vertex_buffers(&mut self, buffer: pso::VertexBufferSet<Backend>) {
-        self.inner().vertex_buffers.clear();
-        self.inner().vertex_buffers.extend(buffer.0.iter().map(|&(buffer, offset)| (buffer.0, offset)));
+    fn bind_vertex_buffers(&mut self, buffer_set: pso::VertexBufferSet<Backend>) {
+        let inner = self.inner();
+        let buffers = &mut inner.resources_vs.buffers;
+        while buffers.len() < buffer_set.0.len()    {
+            buffers.push(None)
+        }
+        for (out, &(buffer, offset)) in buffers.iter_mut().zip(buffer_set.0.iter()) {
+            *out = Some((buffer.0, offset));
+        }
+        if let EncoderState::Render(ref encoder) = inner.encoder_state {
+            for (i, &(buffer, offset)) in buffer_set.0.iter().enumerate() {
+                encoder.set_vertex_buffer(i as _, offset as _, buffer.0);
+            }
+        }
     }
 
     fn set_viewports(&mut self, rects: &[Viewport]) {
+        let inner = self.inner();
         if rects.len() != 1 {
             panic!("Metal supports only one viewport");
         }
         let rect = &rects[0];
-        self.inner().viewport = Some(MTLViewport {
+        let vp = MTLViewport {
             originX: rect.x as f64,
             originY: rect.y as f64,
             width: rect.w as f64,
             height: rect.h as f64,
             znear: rect.near as f64,
             zfar: rect.far as f64,
-        });
+        };
+        inner.viewport = Some(vp);
+        if let EncoderState::Render(ref encoder) = inner.encoder_state {
```

Comment:
Yes, they will be reset. That's why we still fill the cache, so that when we exit a render pass and then enter another one, we can re-bind everything that is active.